lib/ASRouterTargeting.jsm

```diff
@@ -23,31 +23,43 @@ const FRECENT_SITES_IGNORE_BLOCKED = true;
 const FRECENT_SITES_NUM_ITEMS = 25;
 const FRECENT_SITES_MIN_FRECENCY = 100;
 
-const TopFrecentSitesCache = {
-  _lastUpdated: 0,
-  _topFrecentSites: null,
-  get topFrecentSites() {
-    return new Promise(async resolve => {
+function CachedTargetingGetter(property, options = null, updateInterval = FRECENT_SITES_UPDATE_INTERVAL) {
+  const targetingGetter = {
+    _lastUpdated: 0,
+    _value: null,
+    // For testing
+    expire() {
+      this._lastUpdated = 0;
+      this._value = null;
+    }
+  };
+
+  Object.defineProperty(targetingGetter, property, {
+    get: () => new Promise(async resolve => {
       const now = Date.now();
-      if (now - this._lastUpdated >= FRECENT_SITES_UPDATE_INTERVAL) {
-        this._topFrecentSites = await asProvider.getTopFrecentSites({
-          ignoreBlocked: FRECENT_SITES_IGNORE_BLOCKED,
-          numItems: FRECENT_SITES_NUM_ITEMS,
-          topsiteFrecency: FRECENT_SITES_MIN_FRECENCY,
-          onePerDomain: true,
-          includeFavicon: false
-        });
-        this._lastUpdated = now;
+      if (now - targetingGetter._lastUpdated >= updateInterval) {
+        targetingGetter._value = await asProvider[property](options);
```

Comment:
This could use some error handling (i.e. reject if this fails)

Reply:
Also some tests for the cache invalidation behaviour, although a bit hard to do with mochi tests (it would be better with sinon)